### ISSUE TITLE
BUG: Late binding for concat function will use cached value when json input is String

### ISSUE THREAD

**Issue by davsclaus** [json-path/JsonPath#680]
I have been tracking down a bug in camel-jsonpath, that leads me to here
https://github.com/json-path/JsonPath/pull/335/files

In this unit test
https://github.com/json-path/JsonPath/blob/317e68c4ebdd9ce3a62e52b16ef588e503d43aca/json-path/src/test/java/com/jayway/jsonpath/internal/function/Issue234.java

If you change the test to use String as input instead of the Map it fails:

```
    @Test
    public void testIssue234String() {
        String json = "{ \"key\": \"first\"}";
        Object value = JsonPath.read(json, "concat(\"/\", $.key)");
        System.out.println(value);
        json = "{ \"key\": \"second\"}";
        value = JsonPath.read(json, "concat(\"/\", $.key)");
        System.out.println(value);
    }
```

And the output is
```
/first
/first
```

If you turn off cache then it works
```
    @Test
    public void testIssue234String() {
        CacheProvider.setCache(new com.jayway.jsonpath.spi.cache.NOOPCache());
        String json = "{ \"key\": \"first\"}";
        Object value = JsonPath.read(json, "concat(\"/\", $.key)");
        System.out.println(value);
        json = "{ \"key\": \"second\"}";
        value = JsonPath.read(json, "concat(\"/\", $.key)");
        System.out.println(value);
    }
```

And now the output is correct
```
/first
/second
```

**Comment by davsclaus** [comment]
Using jsonpath 2.4.0

**Comment by davsclaus** [comment]
There is also a bug in the original test-case with Map as input. It only works because its the **same** map instance being used, which gets overridden with new data

This test
```
    @Test
    public void testIssue234() {
        Map<String, String> context = new HashMap<>();
        context.put("key", "first");
        Map<String, String> context2 = new HashMap<>();
        context2.put("key", "second2");

        Object value = JsonPath.read(context, "concat(\"/\", $.key)");
        System.out.println(value);
//        context.put("key", "second");
        value = JsonPath.read(context2, "concat(\"/\", $.key)");
        System.out.println(value);
    }
```

Will output wrongly
```
/first
/first
```

**Comment by davsclaus** [comment]
The issue reported at Apache Camel
https://issues.apache.org/jira/browse/CAMEL-16393

**Comment by Pigdrum** [comment]
Hey, I would like to work on this issue. @davsclaus

**Comment by kallestenflo** [comment]
Fixed by #704
